File: ScanAPIDemo/Form1.cs
using System;
using System.Globalization;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Imaging;
using System.Text;
using System.Windows.Forms;
using ScanAPIHelper;
using System.Threading;
using System.Runtime.InteropServices;
using System.IO;

namespace ScanAPIDemo
{
    public partial class Form1 : Form
    {
        delegate void SetTextCallback(string text);

        const int FTR_ERROR_EMPTY_FRAME = 4306; /* ERROR_EMPTY */
        const int FTR_ERROR_MOVABLE_FINGER = 0x20000001;
        const int FTR_ERROR_NO_FRAME = 0x20000002;
        const int FTR_ERROR_USER_CANCELED = 0x20000003;
        const int FTR_ERROR_HARDWARE_INCOMPATIBLE = 0x20000004;
        const int FTR_ERROR_FIRMWARE_INCOMPATIBLE = 0x20000005;
        const int FTR_ERROR_INVALID_AUTHORIZATION_CODE = 0x20000006;

        /* Other return codes are Windows-compatible */
        const int ERROR_NO_MORE_ITEMS = 259;                // ERROR_NO_MORE_ITEMS
        const int ERROR_NOT_ENOUGH_MEMORY = 8;              // ERROR_NOT_ENOUGH_MEMORY
        const int ERROR_NO_SYSTEM_RESOURCES = 1450;         // ERROR_NO_SYSTEM_RESOURCES
        const int ERROR_TIMEOUT = 1460;                     // ERROR_TIMEOUT
        const int ERROR_NOT_READY = 21;                     // ERROR_NOT_READY
        const int ERROR_BAD_CONFIGURATION = 1610;           // ERROR_BAD_CONFIGURATION
        const int ERROR_INVALID_PARAMETER = 87;             // ERROR_INVALID_PARAMETER
        const int ERROR_CALL_NOT_IMPLEMENTED = 120;         // ERROR_CALL_NOT_IMPLEMENTED
        const int ERROR_NOT_SUPPORTED = 50;                 // ERROR_NOT_SUPPORTED
        const int ERROR_WRITE_PROTECT = 19;                 // ERROR_WRITE_PROTECT
        const int ERROR_MESSAGE_EXCEEDS_MAX_SIZE = 4336;    // ERROR_MESSAGE_EXCEEDS_MAX_SIZE

        private Device m_hDevice;
        private bool m_bCancelOperation;
        private byte[] m_Frame;
        private bool m_bScanning;
        private byte m_ScanMode;
        private bool m_bIsLFDSupported;
        private bool m_bExit;

        class ComboBoxItem
        {
            public ComboBoxItem(String value, int interfaceNumber)
            {
                m_String = value;
                m_InterfaceNumber = interfaceNumber;
            }

            public override string ToString()
            {
                return m_String;
            }

            public int interfaceNumber
            {
                get
                {
                    return m_InterfaceNumber;
                }
            }

            private String m_String;
            private int m_InterfaceNumber;
        }

        public Form1()
        {
            InitializeComponent();
            m_btnOpenDevice.Enabled = false;
            m_btnClose.Enabled = false;
            m_grpParameters.Enabled = false;
            m_grpTests.Enabled = false;
            m_hDevice = null;
            m_ScanMode = 0;
            m_bScanning = false;
            m_bExit = false;
        }

        private void OnFormLoad(object sender, EventArgs e)
        {
            try
            {
                int defaultInterface = ScanAPIHelper.Device.BaseInterface;
                FTRSCAN_INTERFACE_STATUS[] status = ScanAPIHelper.Device.GetInterfaces();
                for (int interfaceNumber = 0; interfaceNumber < status.Length; interfaceNumber++)
                {
                    if (status[interfaceNumber] == FTRSCAN_INTERFACE_STATUS.FTRSCAN_INTERFACE_STATUS_CONNECTED)
                    {
                        int index = m_cmbInterfaces.Items.Add( new ComboBoxItem( interfaceNumber.ToString(), interfaceNumber ) );
                        if (defaultInterface == interfaceNumber)
                        {
                            m_cmbInterfaces.SelectedIndex = index;
                        }
                    }
                }
            }
            catch( ScanAPIException ex)
            {
                ShowError( ex );
            }
            m_cmbDose.SelectedIndex = 3;
            m_ScanMode = 0;
        }

        private void m_cmbInterfaces_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (m_cmbInterfaces.SelectedIndex != -1)
            {
                ComboBoxItem item = (ComboBoxItem)m_cmbInterfaces.Items[ m_cmbInterfaces.SelectedIndex ];
                try
                {
                    ScanAPIHelper.Device.BaseInterface = item.interfaceNumber;
                    m_btnOpenDevice.Enabled = true;
                }
                catch (ScanAPIException ex)
                {
                    ShowError(ex);
                }
            }
            else
            {
                m_btnOpenDevice.Enabled = false;
            }

        }

        private void ShowError(ScanAPIException ex)
        {
            string szMessage;
            switch( ex.ErrorCode )
            {
            case FTR_ERROR_EMPTY_FRAME:
                szMessage = "Empty Frame";
                break;

            case FTR_ERROR_MOVABLE_FINGER:
                szMessage = "Movable Finger";
                break;

            case FTR_ERROR_NO_FRAME:
                szMessage = "Fake Finger";
                break;

            case FTR_ERROR_HARDWARE_INCOMPATIBLE:
                szMessage = "Incompatible Hardware";
                break;

            case FTR_ERROR_FIRMWARE_INCOMPATIBLE:
                szMessage = "Incompatible Firmware";
                break;

            case FTR_ERROR_INVALID_AUTHORIZATION_CODE:
                szMessage = "Invalid Authorization Code";
                break;

            case ERROR_NOT_ENOUGH_MEMORY:
                szMessage = "Error code ERROR_NOT_ENOUGH_MEMORY";
                break;

            case ERROR_NO_SYSTEM_RESOURCES:
                szMessage = "Error code ERROR_NO_SYSTEM_RESOURCES";
                break;

            case ERROR_TIMEOUT:
                szMessage = "Error code ERROR_TIMEOUT";
                break;

            case ERROR_NOT_READY:
                szMessage = "Error code ERROR_NOT_READY";
                break;

            case ERROR_BAD_CONFIGURATION:
                szMessage = "Error code ERROR_BAD_CONFIGURATION";
                break;

            case ERROR_INVALID_PARAMETER:
                szMessage = "Error code ERROR_INVALID_PARAMETER";
                break;

            case ERROR_CALL_NOT_IMPLEMENTED:
                szMessage = "Error code ERROR_CALL_NOT_IMPLEMENTED";
                break;

            case ERROR_NOT_SUPPORTED:
                szMessage = "Error code ERROR_NOT_SUPPORTED";
                break;

            case ERROR_WRITE_PROTECT:
                szMessage = "Error code ERROR_WRITE_PROTECT";
                break;

            case ERROR_MESSAGE_EXCEEDS_MAX_SIZE:
                szMessage = "Error code ERROR_MESSAGE_EXCEEDS_MAX_SIZE";
                break;

            default:
                szMessage = String.Format( "Error code: {0}", ex.ErrorCode );
                break;
            }
            SetMessageText( szMessage );
        }

        private void OnOpenDevice(object sender, EventArgs e)
        {
            Console.WriteLine("hello");
            try
            {
                m_hDevice = new Device();
                m_hDevice.Open();

                // gets devce parameters
                VersionInfo version = m_hDevice.VersionInformation;
                m_lblApiVersion.Text = version.APIVersion.ToString();
                m_lblHardwareVersion.Text = version.HardwareVersion.ToString();
                m_lblFirmwareVersion.Text = version.FirmwareVersion.ToString();

                m_chkDetectFakeFinger.Checked = m_hDevice.DetectFakeFinger;

                m_bIsLFDSupported = false;
                DeviceInfo dinfo = m_hDevice.Information;
                switch (dinfo.DeviceCompatibility)
                {
                    case 0:
                    case 1:
                    case 4:
                    case 11:
                        m_lblCompatibility.Text = "FS80";
                        m_bIsLFDSupported = true;
                        break;
                    case 5:
                        m_lblCompatibility.Text = "FS88";
                        m_bIsLFDSupported = true;
                        break;
                    case 7:
                        m_lblCompatibility.Text = "FS50";
                        break;
                    case 8:
                        m_lblCompatibility.Text = "FS60";
                        break;
                    case 9:
                        m_lblCompatibility.Text = "FS25";
                        m_bIsLFDSupported = true;
                        break;
                    case 10:
                        m_lblCompatibility.Text = "FS10";
                        break;
                    case 13:
                        m_lblCompatibility.Text = "FS80H";
                        m_bIsLFDSupported = true;
                        break;
                    case 14:
                        m_lblCompatibility.Text = "FS88H";
                        m_bIsLFDSupported = true;
                        break;
                    case 15:
                        m_lblCompatibility.Text = "FS64";
                        break;
                    case 16:
                        m_lblCompatibility.Text = "FS26E";
                        break;
                    case 17:
                        m_lblCompatibility.Text = "FS80HS";
                        break;
                    case 18:
                        m_lblCompatibility.Text = "FS26";
                        break;
                    default:
                        m_lblCompatibility.Text = "Unknown device";
                        break;
                }

                m_lblCurrentImageSize.Text = m_hDevice.ImageSize.ToString();

                m_grpParameters.Enabled = true;
                m_grpTests.Enabled = true;

                m_cmbInterfaces.Enabled = false;
                m_btnOpenDevice.Enabled = false;
                m_btnClose.Enabled = true;                
                m_chkDetectFakeFinger.Enabled = m_bIsLFDSupported;
            }
            catch( ScanAPIException ex )
            {
                if( m_hDevice != null )
                {
                    m_hDevice.Dispose();
                    m_hDevice = null;
                }
                ShowError(ex);
            }

        }

        private void OnDetectFakeFinger(object sender, EventArgs e)
        {
            if (m_hDevice != null)
            {
                m_hDevice.DetectFakeFinger = m_chkDetectFakeFinger.Checked;
            }
        }

        private void OnCloseDevice(object sender, EventArgs e)
        {
            m_bCancelOperation = true;
            Size size = m_hDevice.ImageSize;
            m_hDevice.Dispose();
            m_hDevice = null;

            m_lblApiVersion.Text = String.Empty;
            m_lblHardwareVersion.Text = String.Empty;
            m_lblFirmwareVersion.Text = String.Empty;

            m_chkDetectFakeFinger.Checked = false;

            m_lblCompatibility.Text = String.Empty;

            m_lblCurrentImageSize.Text = String.Empty;

            m_picture.Image = null;

            m_grpParameters.Enabled = false;
            m_grpTests.Enabled = false;

            m_cmbInterfaces.Enabled = true;
            m_btnOpenDevice.Enabled = true;
            m_btnClose.Enabled = false;
            m_btnSave.Enabled = false;
        }

        private void SetMessageText(string text)
        {
            if (m_bExit)
                return;
            if (this.m_textMessage.InvokeRequired)
            {
                SetTextCallback d = new SetTextCallback(this.SetMessageText);
                this.Invoke(d, new object[] { text });
            }
            else
            {
                this.m_textMessage.Text = text;
                this.Update();
            }
        }

        [STAThread]
        private void GetFrame()
        {
            try
            {
                if (m_ScanMode == 0)
                    m_Frame = m_hDevice.GetFrame();
                else
                    m_Frame = m_hDevice.GetImage(m_ScanMode);
                SetMessageText("OK");
                Thread myth;
                myth = new Thread(new System.Threading.ThreadStart(CallSaveDialog));
                myth.ApartmentState = ApartmentState.STA;
                myth.Start();

            }
            catch (ScanAPIException ex)
            {
                if (m_Frame != null)
                    m_Frame = null;
                ShowError(ex);
            }
        }

        private void CallSaveDialog()
        {
            SaveFileDialog dlgSave = new SaveFileDialog();
            dlgSave.Filter = "bmp files (*.bmp)|*.bmp|wsq files (*.wsq)|*.wsq";
            if (dlgSave.ShowDialog() == DialogResult.OK)
            {
                if (dlgSave.FilterIndex == 1)
                {
                    MyBitmapFile myFile = new MyBitmapFile(m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, m_Frame);
                    FileStream file = new FileStream(dlgSave.FileName, FileMode.Create);
                    file.Write(myFile.BitmatFileData, 0, myFile.BitmatFileData.Length);
                    file.Close();
                    SetMessageText("Bitmap file is saved to " + dlgSave.FileName);
                }
                else //wsq
                {
                    float fBitRate = 0.75f; // in the range of 0.75 - 2.25, lower value with higher compression rate
                    byte[] wsqImage = m_hDevice.WSQ_FromRawImage(m_Frame, m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, fBitRate);
                    if (wsqImage != null)
                    {
                        FileStream file = new FileStream(dlgSave.FileName, FileMode.Create);
                        file.Write(wsqImage, 0, wsqImage.Length);
                        file.Close();
                        SetMessageText("WSQ file is saved to " + dlgSave.FileName);
                    }
                }
            }

        }

        private void OnTestGetFrame(object sender, EventArgs e)
        {
            if (!m_bScanning)
            {
                m_bCancelOperation = false;
                m_btnGetFrame.Text = "Stop";
                m_btnSave.Enabled = false;
                m_btnClose.Enabled = false;
                Thread WorkerThread = new Thread(new ThreadStart(CaptureThread));
                WorkerThread.Start();
            }
            else
            {
                m_bCancelOperation = true;
                m_btnGetFrame.Text = "Scan";
                m_btnClose.Enabled = true;
                if (m_Frame != null)
                    m_btnSave.Enabled = true;
            }
        }

        private void CaptureThread()
        {
            m_bScanning = true;
            while (!m_bCancelOperation)
            {
                GetFrame();
                if (m_Frame != null)
                {
                    MyBitmapFile myFile = new MyBitmapFile(m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, m_Frame);
                    MemoryStream BmpStream = new MemoryStream(myFile.BitmatFileData);
                    Bitmap Bmp = new Bitmap(BmpStream);
                    m_picture.Image = Bmp;
                }
                else
                    m_picture.Image = null;
                Thread.Sleep(10);
            }
            m_bScanning = false;
        }

        private void OnFormClosed(object sender, FormClosedEventArgs e)
        {
            m_bExit = true;
            m_bCancelOperation = true;
            if (m_hDevice != null)
            {
                m_hDevice.Dispose();
                m_hDevice = null;
            }
        }

        private void m_btnSave_Click(object sender, EventArgs e)
        {
            SaveFileDialog dlgSave = new SaveFileDialog();
            dlgSave.Filter = "bmp files (*.bmp)|*.bmp|wsq files (*.wsq)|*.wsq";
            if (dlgSave.ShowDialog() == DialogResult.OK)
            {
                if (dlgSave.FilterIndex == 1)
                {
                    MyBitmapFile myFile = new MyBitmapFile(m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, m_Frame);
                    FileStream file = new FileStream(dlgSave.FileName, FileMode.Create);
                    file.Write(myFile.BitmatFileData, 0, myFile.BitmatFileData.Length);
                    file.Close();
                    SetMessageText("Bitmap file is saved to " + dlgSave.FileName);
                }
                else //wsq
                {
                    float fBitRate = 0.75f; // in the range of 0.75 - 2.25, lower value with higher compression rate
                    byte[] wsqImage = m_hDevice.WSQ_FromRawImage(m_Frame, m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, fBitRate);
                    if (wsqImage != null)
                    {
                        FileStream file = new FileStream(dlgSave.FileName, FileMode.Create);
                        file.Write(wsqImage, 0, wsqImage.Length);
                        file.Close();
                        SetMessageText("WSQ file is saved to " + dlgSave.FileName);
                    }                        
                }
            }
        }

        private void m_chkInvertImage_CheckedChanged(object sender, EventArgs e)
        {
            if (m_hDevice != null)
            {
                m_hDevice.InvertImage = m_chkInvertImage.Checked;
            }
        }

        private void m_chkFrame_CheckedChanged(object sender, EventArgs e)
        {
            if (m_chkFrame.Checked)
                m_ScanMode = 0;
            else
                m_ScanMode = (byte)(m_cmbDose.SelectedIndex + 1);
            m_cmbDose.Enabled = (!m_chkFrame.Checked);
            m_chkDetectFakeFinger.Enabled = (m_bIsLFDSupported ? m_chkFrame.Checked : false );
        }

        private void m_cmbDose_SelectedIndexChanged(object sender, EventArgs e)
        {
            m_ScanMode = (byte) (m_cmbDose.SelectedIndex + 1);
        }
    }
}

File: ScanAPIDemo_VB/Form1.vb
Imports ScanAPIHelper
Imports System.Threading
Imports System.IO


Public Class Form1

    delegate sub SetTextCallback(text as String)

    Public Const FTR_ERROR_EMPTY_FRAME As Integer = 4306
    Public Const FTR_ERROR_MOVABLE_FINGER As Integer = &H20000001
    Public Const FTR_ERROR_NO_FRAME As Integer = &H20000002
    Public Const FTR_ERROR_USER_CANCELED As Integer = &H20000003
    Public Const FTR_ERROR_HARDWARE_INCOMPATIBLE As Integer = &H20000004
    Public Const FTR_ERROR_FIRMWARE_INCOMPATIBLE As Integer = &H20000005
    Public Const FTR_ERROR_INVALID_AUTHORIZATION_CODE As Integer = &H20000006

    Private m_hDevice As Device
    Private m_bCancelOperation As Boolean
    Private m_Frame() As Byte
    Private m_bScanning As Boolean
    Private m_ScanMode As Byte
    Private m_bIsLFDSupported As Boolean
    Private m_bExit As Boolean

    Public Class ComboBoxItem
        Dim m_String As String
        Dim m_InterfaceNumber As Integer

        Public Sub New(ByVal value As String, ByVal interfaceNumber As Integer)
            m_String = value
            m_InterfaceNumber = interfaceNumber
        End Sub

        Public Overrides Function ToString() As String
            Return m_String
        End Function

        Public ReadOnly Property interfaceNumber() As Integer
            Get
                Return m_InterfaceNumber
            End Get
        End Property
    End Class


    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        m_btnOpenDevice.Enabled = False
        m_btnClose.Enabled = False
        m_grpParameters.Enabled = False
        m_grpTests.Enabled = False
        m_hDevice = Nothing
        m_ScanMode = 0
        m_bScanning = False
        m_bExit = False
        Dim defaultInterface As Integer
        Dim status() As FTRSCAN_INTERFACE_STATUS

        Try
            defaultInterface = ScanAPIHelper.Device.BaseInterface
            status = ScanAPIHelper.Device.GetInterfaces()
            For i As Integer = 0 To status.Length - 1
                If status(i) = FTRSCAN_INTERFACE_STATUS.FTRSCAN_INTERFACE_STATUS_CONNECTED Then
                    Dim index As Integer
                    index = m_cmbInterfaces.Items.Add(New ComboBoxItem(i.ToString(), i))
                    If (defaultInterface = i) Then
                        m_cmbInterfaces.SelectedIndex = index
                    End If
                End If
            Next i
        Catch ex As ScanAPIException
            ShowError(ex)
        End Try
        m_cmbDose.SelectedIndex = 3
        m_ScanMode = 0
    End Sub

    Private Sub m_btnOpenDevice_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_btnOpenDevice.Click
        Try
            m_hDevice = New Device()
            m_hDevice.Open()
            ' gets devce parameters
            Dim version As VersionInfo = m_hDevice.VersionInformation
            m_lblApiVersion.Text = version.APIVersion.ToString()
            m_lblHardwareVersion.Text = version.HardwareVersion.ToString()
            m_lblFirmwareVersion.Text = version.FirmwareVersion.ToString()

            m_chkDetectFakeFinger.Checked = m_hDevice.DetectFakeFinger

            m_bIsLFDSupported = False
            Dim dinfo As DeviceInfo = m_hDevice.Information
            Select Case (dinfo.DeviceCompatibility)
                Case 0
                    m_lblCompatibility.Text = "FS80"
                    m_bIsLFDSupported = True
                Case 1
                    m_lblCompatibility.Text = "FS80"
                    m_bIsLFDSupported = True
                Case 4
                    m_lblCompatibility.Text = "FS80"
                    m_bIsLFDSupported = True
                Case 11
                    m_lblCompatibility.Text = "FS80"
                    m_bIsLFDSupported = True
                Case 5
                    m_lblCompatibility.Text = "FS88"
                    m_bIsLFDSupported = True
                Case 7
                    m_lblCompatibility.Text = "FS50"
                Case 8
                    m_lblCompatibility.Text = "FS60"
                Case 9
                    m_lblCompatibility.Text = "FS25"
                    m_bIsLFDSupported = True
                Case 10
                    m_lblCompatibility.Text = "FS10"
                Case 13
                    m_lblCompatibility.Text = "FS80H"
                    m_bIsLFDSupported = True
                Case 14
                    m_lblCompatibility.Text = "FS88H"
                    m_bIsLFDSupported = True
                Case 15
                    m_lblCompatibility.Text = "FS64"
                Case 16
                    m_lblCompatibility.Text = "FS26E"
                Case 17
                    m_lblCompatibility.Text = "FS80HS"
                Case 18
                    m_lblCompatibility.Text = "FS26"
                Case Else
                    m_lblCompatibility.Text = "Unknown device"
            End Select

            m_lblCurrentImageSize.Text = m_hDevice.ImageSize.ToString()
            m_grpParameters.Enabled = True
            m_grpTests.Enabled = True

            m_cmbInterfaces.Enabled = False
            m_btnOpenDevice.Enabled = False
            m_btnClose.Enabled = True
            m_chkDetectFakeFinger.Enabled = m_bIsLFDSupported
        Catch ex As ScanAPIException
            If m_hDevice IsNot Nothing Then
                m_hDevice.Dispose()
                m_hDevice = Nothing
                ShowError(ex)
            End If
        End Try
    End Sub

    Private Sub m_btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_btnClose.Click
        m_bCancelOperation = True
        m_hDevice.Dispose()
        m_hDevice = Nothing

        m_lblApiVersion.Text = String.Empty
        m_lblHardwareVersion.Text = String.Empty
        m_lblFirmwareVersion.Text = String.Empty

        m_chkDetectFakeFinger.Checked = False

        m_lblCompatibility.Text = String.Empty

        m_lblCurrentImageSize.Text = String.Empty

        m_picture.Image = Nothing

        m_grpParameters.Enabled = False
        m_grpTests.Enabled = False

        m_cmbInterfaces.Enabled = True
        m_btnOpenDevice.Enabled = True
        m_btnClose.Enabled = False
        m_btnSave.Enabled = False

    End Sub

    Private Sub m_cmbInterfaces_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_cmbInterfaces.SelectedIndexChanged
        If m_cmbInterfaces.SelectedIndex <> -1 Then
            Dim item As ComboBoxItem = m_cmbInterfaces.Items(m_cmbInterfaces.SelectedIndex)
            Try
                ScanAPIHelper.Device.BaseInterface = item.interfaceNumber
                m_btnOpenDevice.Enabled = True
            Catch ex As ScanAPIException
                ShowError(ex)
            End Try
        Else
            m_btnOpenDevice.Enabled = False
        End If
    End Sub

    Private Sub ShowError(ByRef ex As ScanAPIException)
        Dim szMessage As String
        Select Case (ex.ErrorCode)
            Case FTR_ERROR_EMPTY_FRAME
                szMessage = "Empty Frame"
            Case FTR_ERROR_MOVABLE_FINGER
                szMessage = "Movable Finger"
            Case FTR_ERROR_NO_FRAME
                szMessage = "Fake Finger"
            Case FTR_ERROR_HARDWARE_INCOMPATIBLE
                szMessage = "Incompatible Hardware"
            Case FTR_ERROR_FIRMWARE_INCOMPATIBLE
                szMessage = "Incompatible Firmware"
            Case FTR_ERROR_INVALID_AUTHORIZATION_CODE
                szMessage = "Invalid Authorization Code"
            Case Else
                szMessage = String.Format("Error code: {0}", ex.ErrorCode)
        End Select
        SetMessageText(szMessage)
    End Sub

    Private Sub SetMessageText(ByVal text As String)
        If m_bExit Then
            Exit Sub
        End If
        If (m_textMessage.InvokeRequired) Then
            Dim d As SetTextCallback = New SetTextCallback(AddressOf Me.SetMessageText)
            Me.Invoke(d, New Object() {text})
        Else
            Me.m_textMessage.Text = text
            Me.Update()
        End If
    End Sub

    Private Sub m_btnGetFrame_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_btnGetFrame.Click
        If Not m_bScanning Then
            m_bCancelOperation = False
            m_btnGetFrame.Text = "Stop"
            m_btnSave.Enabled = False
            m_btnClose.Enabled = False
            Dim WorkerThread As Thread = New Thread(New ThreadStart(AddressOf CaptureThread))
            WorkerThread.Start()
        Else
            m_bCancelOperation = True
            m_btnGetFrame.Text = "Scan"
            m_btnClose.Enabled = True
            If m_Frame IsNot Nothing Then
                m_btnSave.Enabled = True
            End If
        End If
    End Sub

    Private Sub CaptureThread()
        m_bScanning = True
        While (Not m_bCancelOperation)
            GetFrame()
            If m_Frame IsNot Nothing Then
                Dim myFile As MyBitmapFile = New MyBitmapFile(m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, m_Frame)
                Dim BmpStream As MemoryStream = New MemoryStream(myFile.BitmatFileData)
                Dim Bmp As Bitmap = New Bitmap(BmpStream)
                m_picture.Image = Bmp
            Else
                m_picture.Image = Nothing
            End If
            Thread.Sleep(10)
        End While
        m_bScanning = False
    End Sub

    Private Sub GetFrame()
        Try
            If m_ScanMode = 0 Then
                m_Frame = m_hDevice.GetFrame()
            Else
                m_Frame = m_hDevice.GetImage(m_ScanMode)
            End If
            SetMessageText("OK")
        Catch ex As ScanAPIException
            If m_Frame IsNot Nothing Then
                m_Frame = Nothing
                ShowError(ex)
            End If
        End Try
    End Sub


    Private Sub m_chkFrame_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_chkFrame.CheckedChanged
        If (m_chkFrame.Checked) Then
            m_ScanMode = 0
        Else
            m_ScanMode = m_cmbDose.SelectedIndex + 1
        End If
        m_cmbDose.Enabled = Not m_chkFrame.Checked
        If m_bIsLFDSupported Then
            m_chkDetectFakeFinger.Enabled = m_chkFrame.Checked
        Else
            m_chkDetectFakeFinger.Enabled = False
        End If
    End Sub

    Private Sub m_cmbDose_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_cmbDose.SelectedIndexChanged
        m_ScanMode = m_cmbDose.SelectedIndex + 1
    End Sub

    Private Sub m_chkDetectFakeFinger_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_chkDetectFakeFinger.CheckedChanged
        If m_hDevice IsNot Nothing Then
            m_hDevice.DetectFakeFinger = m_chkDetectFakeFinger.Checked
        End If
    End Sub

    Private Sub m_chkInvertImage_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_chkInvertImage.CheckedChanged
        If m_hDevice IsNot Nothing Then
            m_hDevice.InvertImage = m_chkInvertImage.Checked
        End If
    End Sub

    Private Sub m_btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles m_btnSave.Click
        Dim dlgSave As SaveFileDialog = New SaveFileDialog()
        dlgSave.Filter = "bmp files (*.bmp)|*.bmp|wsq files (*.wsq)|*.wsq"
        If dlgSave.ShowDialog() = DialogResult.OK Then
            If dlgSave.FilterIndex = 1 Then
                Dim myFile As MyBitmapFile = New MyBitmapFile(m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, m_Frame)
                Dim file As FileStream = New FileStream(dlgSave.FileName, FileMode.Create)
                file.Write(myFile.BitmatFileData, 0, myFile.BitmatFileData.Length)
                file.Close()
                SetMessageText("Bitmap file is saved to " + dlgSave.FileName)
            Else
                Dim fBitRate As Single = 0.75  ' in the range of 0.75 - 2.25, lower value with higher compression rate
                Dim wsqImage() As Byte = m_hDevice.WSQ_FromRawImage(m_Frame, m_hDevice.ImageSize.Width, m_hDevice.ImageSize.Height, fBitRate)
                If wsqImage IsNot Nothing Then
                    Dim file As FileStream = New FileStream(dlgSave.FileName, FileMode.Create)
                    file.Write(wsqImage, 0, wsqImage.Length)
                    file.Close()
                    SetMessageText("WSQ file is saved to " + dlgSave.FileName)
                End If
            End If
        End If
    End Sub

    Private Sub Form1_FormClosed(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed
        m_bExit = True
        m_bCancelOperation = True
        If m_hDevice IsNot Nothing Then
            m_hDevice.Dispose()
            m_hDevice = Nothing
        End If
    End Sub
End Class
